src/reactivity.ts

```diff
@@ -1,277 +1,163 @@
-        delete target[key];
-        // notify source observers
-        const clean = isTrackable(current);
-        notify(sourceAtoms.get(source)!.get(ROOT)!, clean);
-        const atoms = sourceAtoms.get(source)!;
-        if (atoms.has(key)) {
-          // clear source-key observers
-          for (const atom of atoms.get(key)!) {
-            atom[KEYS].delete(key);
-          }
-          atoms.delete(key);
-        }
-      }
-      return true;
-    },
-    get(target: any, key: any, proxy: any): any {
-      switch (key) {
-        case OBSERVER:
-          return observer;
-        case SOURCE:
-          return source;
-        case KEYS:
-          return keys;
-        case SEED:
-          return seed;
-        default:
-          const value = target[key];
-          // register observer to source-key
-          if (!keys.has(key) && observerSourceAtom.has(observer)) {
-            const atoms = sourceAtoms.get(source)!;
-            if (!atoms.has(key)) {
-              atoms.set(key, new ObserverSet());
-            }
-            atoms.get(key)!.add(newAtom);
-            keys.add(key);
-          }
-          //
-          return _atom(value, observer);
-      }
-    },
-  });
-  return newAtom;
-}
-
-function isTrackable(value: any): boolean {
+function isObservable(value: any) {
   return (
-    value !== null &&
     typeof value === "object" &&
+    value !== null &&
     !(value instanceof Date) &&
     !(value instanceof Promise) &&
     !(value instanceof String) &&
     !(value instanceof Number)
   );
 }
 
-function registerObserver(observer: Observer) {
-  if (!observerSourceAtom.get(observer)) {
-    observerSourceAtom.set(observer, new Map());
+const targetToKeysToCallbacks = new WeakMap<Target, Map<ObjectKey, Set<callback>>>();
+function observeTargetKey(target: Target, key: ObjectKey, callback: callback) {
+  if (!targetToKeysToCallbacks.get(target)) {
+    targetToKeysToCallbacks.set(target, new Map());
+  }
+  const keyToCallbacks = targetToKeysToCallbacks.get(target)!;
+  if (!keyToCallbacks.get(key)) {
+    keyToCallbacks.set(key, new Set());
   }
-  return unregisterObserverAtoms.bind(null, observer, false);
+  keyToCallbacks.get(key)!.add(callback);
 }
 
-function unregisterObserverAtoms(observer: Observer, keepSeeds: boolean) {
-  const observerAtoms = observerSourceAtom.get(observer)!;
-  for (const [source, atom] of observerAtoms) {
-    if (keepSeeds && atom[SEED]) {
-      continue;
-    }
-    observerAtoms.delete(source);
-    const atoms = sourceAtoms.get(source)!;
-    atoms.get(ROOT)!.delete(atom);
-    for (const key of atom[KEYS]) {
-      atoms.get(key)!.delete(atom);
-    }
+function notifyTarget(target: Target, key: ObjectKey) {
+  const keyToCallbacks = targetToKeysToCallbacks.get(target);
+  if (!keyToCallbacks) {
+    return;
   }
-  if (!keepSeeds) {
-    observerSourceAtom.delete(observer);
+  const callbacks = [...(keyToCallbacks.get(key) || [])];
```

Comment:
why not callbacks = keyToCallbacks.get(key)
then return if !callbacks?

Reply:
because `clearReactivesToCallbacks` will modify the set in place and mess with the iteration on it just below, so we need to loop on a copy of the entries.